pkg/node/node.go

```diff
@@ -1665,89 +1635,504 @@ func (s *Service) buildInitiatorsArray(initiators []string, arrayID string) []go
+			}
+			arrayAddedList[remoteArr.GlobalID] = true
+
+			if remoteConn == gopowerstore.HostConnectivityEnumMetroOptimizeBoth {
+				coLocated = true
+			}
 		}
 	}
+	return coLocated, nil
+}
+
+// isRemoteToOtherArray returns true if arrA and arrB are mutually remote to each other.
+func (s *Service) isRemoteToOtherArray(
+	ctx context.Context,
+	arrA, arrB *array.PowerStoreArray,
+) bool {
+	// fetch arrA’s remotes
+	remotesA, err := arrA.GetClient().GetAllRemoteSystems(ctx)
+	if err != nil {
+		log.Warnf("[isRemoteToOtherArray] failed to get remotes for %s: %v", arrA.GlobalID, err)
+		return false
+	}
+	// fetch arrB’s remotes
+	remotesB, err := arrB.GetClient().GetAllRemoteSystems(ctx)
+	if err != nil {
+		log.Warnf("[isRemoteToOtherArray] failed to get remotes for %s: %v", arrB.GlobalID, err)
+		return false
+	}
+
+	foundAtoB := false
+	for _, r := range remotesA {
+		if r.SerialNumber == arrB.GlobalID {
+			foundAtoB = true
+			break
+		}
+	}
+	if !foundAtoB {
+		return false
+	}
+
+	for _, r := range remotesB {
+		if r.SerialNumber == arrA.GlobalID {
+			return true
+		}
+	}
+	return false
+}
+
+// Checks if host with given initiators already exists
+func (s *Service) isHostAlreadyRegistered(ctx context.Context, client gopowerstore.Client, initiators []string) bool {
+	existingHosts, err := client.GetHosts(ctx)
+	if err != nil {
+		log.Warnf("[isHostAlreadyRegistered] Failed to get hosts: %v", err)
+		return false
+	}
+
+	for _, host := range existingHosts {
+		for _, hInit := range host.Initiators {
+			for _, i := range initiators {
+				if hInit.PortName == i {
+					log.Infof("[isHostAlreadyRegistered] Found existing host with initiator %s", i)
+					return true
+				}
+			}
+		}
+	}
+	return false
+}
+
+func (s *Service) registerHost(
+	ctx context.Context,
+	client gopowerstore.Client,
+	arrayID string,
+	initiators []string,
+	connType gopowerstore.HostConnectivityEnum,
+) error {
+	description := fmt.Sprintf("k8s node: %s", s.opts.KubeNodeName)
+	reqInitiators := s.buildInitiatorsArray(initiators, arrayID)
+	osType := gopowerstore.OSTypeEnumLinux
+
+	createParams := gopowerstore.HostCreate{
+		Name:             &s.nodeID,
+		OsType:           &osType,
+		Initiators:       &reqInitiators,
+		Description:      &description,
+		HostConnectivity: connType,
+	}
+
+	if s.opts.KubeNodeName != "" && common.IsK8sMetadataSupported(client) {
+		metadata := map[string]string{"k8s_node_name": s.opts.KubeNodeName}
+		createParams.Metadata = &metadata
+
+		headers := client.GetCustomHTTPHeaders()
+		if headers == nil {
+			headers = make(http.Header)
+		}
+		headers.Add("DELL-VISIBILITY", "internal")
+		client.SetCustomHTTPHeaders(headers)
+	}
+
+	log.Infof("[registerHost] Creating host on array %s with connectivity: %s", arrayID, connType)
 	resp, err := client.CreateHost(ctx, &createParams)
-	// reset custom header
-	customHeaders.Del("DELL-VISIBILITY")
-	client.SetCustomHTTPHeaders(customHeaders)
+	client.SetCustomHTTPHeaders(nil)
```

Comment:
Shoudn't we just `customHeaders.Del("DELL-VISIBILITY")` i.e. deleting/removing this particular header rather than setting the complete header to  `client.SetCustomHTTPHeaders(nil)` nil?

Reply:
This is fixed